spark/src/main/java/org/apache/iceberg/RemoveOrphanFilesAction.java

```diff
@@ -0,0 +1,239 @@
+    Dataset<Row> manifestDF = spark.read().format("iceberg")
+        .load(allManifestsMetadataTable)
+        .selectExpr("path as file_path");
+
+    List<String> otherMetadataFiles = Lists.newArrayList();
+
+    for (Snapshot snapshot : table.snapshots()) {
+      String manifestListLocation = snapshot.manifestListLocation();
+      if (manifestListLocation != null) {
+        otherMetadataFiles.add(manifestListLocation);
+      }
+    }
+
+    otherMetadataFiles.add(ops.metadataFileLocation("version-hint.text"));
+
+    TableMetadata metadata = ops.current();
+    otherMetadataFiles.add(metadata.file().location());
+    for (TableMetadata.MetadataLogEntry previousMetadataFile : metadata.previousFiles()) {
+      otherMetadataFiles.add(previousMetadataFile.file());
+    }
+
+    Dataset<Row> otherMetadataFileDF = spark
+        .createDataset(otherMetadataFiles, Encoders.STRING())
+        .toDF("file_path");
+
+    return manifestDF.union(otherMetadataFileDF);
+  }
+
+  private Dataset<Row> buildActualFileDF() {
+    List<String> topLevelDirs = Lists.newArrayList();
+    List<String> matchingTopLevelFiles = Lists.newArrayList();
+
+    try {
+      Path path = new Path(location);
+      FileSystem fs = path.getFileSystem(hadoopConf.value());
+
+      for (FileStatus file : fs.listStatus(path, HiddenPathFilter.get())) {
+        if (file.isDirectory()) {
+          topLevelDirs.add(file.getPath().toString());
+        } else if (file.isFile() && file.getModificationTime() < olderThanTimestamp) {
+          matchingTopLevelFiles.add(file.getPath().toString());
+        }
+      }
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to determine top-level files and dirs in {}", location);
+    }
+
+    JavaRDD<String> matchingTopLevelFileRDD = sparkContext.parallelize(matchingTopLevelFiles, 1);
+
+    if (topLevelDirs.isEmpty()) {
+      return spark.createDataset(matchingTopLevelFileRDD.rdd(), Encoders.STRING()).toDF("file_path");
+    }
+
+    int parallelism = Math.min(topLevelDirs.size(), partitionDiscoveryParallelism);
+    JavaRDD<String> topLevelDirRDD = sparkContext.parallelize(topLevelDirs, parallelism);
+
+    Broadcast<SerializableConfiguration> conf = sparkContext.broadcast(hadoopConf);
+    JavaRDD<String> matchingLeafFileRDD = topLevelDirRDD.mapPartitions(listDirsRecursively(conf, olderThanTimestamp));
+
+    JavaRDD<String> matchingFileRDD = matchingTopLevelFileRDD.union(matchingLeafFileRDD);
+    return spark.createDataset(matchingFileRDD.rdd(), Encoders.STRING()).toDF("file_path");
+  }
+
+  private static FlatMapFunction<Iterator<String>, String> listDirsRecursively(
+      Broadcast<SerializableConfiguration> conf,
+      long olderThanTimestamp) {
+
+    return (FlatMapFunction<Iterator<String>, String>) dirs -> {
+      List<String> files = Lists.newArrayList();
+      Predicate<FileStatus> predicate = file -> file.getModificationTime() < olderThanTimestamp;
```

Comment:
Nit: `predicate` isn't a very descriptive name. Maybe `pastOperationTimeLimit` instead?

Reply:
I kept it to stay on one line below.